[Optimization] Use protobuf objects directly instead of dicts

friscoMad opened the pull request:
## Description
This PR does avoid converting all response data from host, ~~to work needs a pgoapi PR (https://github.com/pogodevorg/pgoapi/pull/200) to be merged but after 2 weeks waiting without reply there I think that maybe posting this here could speed up review there.~~
API changes have been merged now so to test it you need to checkout this PR and then:
```
pip uninstall pgoapi
pip install --upgrade -r requirements.txt
```

## Motivation and Context
There is no gain in converting all objects to dicts and there are some issues:

- Wasted CPU time: Responses are converted to protobuf objects and then to dicts, also accessing dicts it is slower than accesing regular object attributes (even if O(1) there is no need of hashing and comparing the key before retrieval)
- Increase memory usage:  Dicts use more memory than regular objects as there is no need to store keys.
- Protobuf2dict has problems with falsy values: There are a lot of code that needs to check against this issue like gym checks in map parsing or empty lists, in the object the value is there the problem is within the conversion.

## How Has This Been Tested?
This code has been working for 3 weeks in my server, but I did not test -tut or encounters so I am almost sure that there is some error in that part as a lot of code needs to be converted and I have probable missed some lines.

## Screenshots (if appropriate):
Memory consumption of a -st 41 280 worker instance with develop (it is test instance):
![pokemap tlp - pokemad neoseo es_22 - bitvise xterm_2017-05-09_14-03-00](https://cloud.githubusercontent.com/assets/487098/25873110/707f5db4-350d-11e7-8bfe-1dfd890a7371.png)

Memory consumption of a -st 41 280 worker instance with this branch (it is test instance):
![pokemap tlp - pokemad neoseo es_22 - bitvise xterm_2017-05-09_18-19-17](https://cloud.githubusercontent.com/assets/487098/25873146/989d49dc-350d-11e7-9538-1653670a9e08.png)

I timed the time spent parsing the response of every request with develop and with this pr against the pgoapi PR (it is backwards compatible). Tested with an instance of -st 5 and 1 thread to reduce cpu load noise and context switching problems:

Develop:
![pokemap tlp - pokemad neoseo es_22 - bitvise xterm_2017-05-09_22-41-18](https://cloud.githubusercontent.com/assets/487098/25873118/7a21aec6-350d-11e7-9b9f-0128335f04db.png)

Protobuf:
![pokemap tlp - pokemad neoseo es_22 - bitvise xterm_2017-05-09_22-33-19](https://cloud.githubusercontent.com/assets/487098/25874173/b10e417a-3511-11e7-94ec-0acacd3aa385.png)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Improvement

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--  NOTE: In order to check code style locally and avoid having your build rejected by Travis, -->
<!--  run the following commands before you commit: `flake8 .` and `npm run lint`. Fix any -->
<!--  issues they point out. Note also that flake's NOQA is disabled on Travis. -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

MARTINI1 commented:
This pr if that is an authentic zoom zoom to optimize the map, you make amazings PRs for optimizing map and not the wild trolls that appear from time to time making PRs useless xD

Almost 200 MB is a great improvement, above all I will note that I use a map in an rpi3, good job ;)

@friscoMad Hmm, I wanted to test it but nothing else to boot and while the workers connect I get this error when connecting each one:

[ERROR] Exception in search_worker under account xxxx Exception message: TypeError ('call () takes exactly 1 argument (2 given)',).

Arengo commented:
@MARTINI1 there is only one real troll @bluemode ;-)

sebastienvercammen commented:
I like it, but we need a reply from pgoapi. I'm on the API maintainers list in pogodev, so I'll poke them for feedback.

michikrug commented:
@friscoMad you probably missed two spots

access to fort properties in:

https://github.com/RocketMap/RocketMap/pull/2026/files#diff-c08aa0b614366984e8695649c57eded7R260
and
https://github.com/RocketMap/RocketMap/pull/2026/files#diff-c08aa0b614366984e8695649c57eded7R283

friscoMad commented:
@michikrug Nice found, I fixed them along with some more in the spin poke request.
Also rebased and working with latest api

michikrug commented:
Encounter also doesn't seem to work, since the get_player_level for the encounter_result gives 0

friscoMad commented:
@michikrug It should be fixed by now at least the get_player_level, thanks for the report

michikrug commented:
Thanks. 
The last remaining error I currently get is:

```
pogom/account.py", line 292, in spin_pokestop
captcha_url = spin_response['responses'][
TypeError: 'bool' object has no attribute '__getitem__'
```

Obihoernchen commented:
https://github.com/pogodevorg/pgoapi/pull/200 is merged now :thumbsup:

friscoMad commented:
@michikrug  I fixed those exceptions but they are only cascade exceptions, when the requests fails the response is a false object and does not work as a dict, so they shouldn't be related with this PR, anyway I am fixing them as it is pointless to add a new PR for just 2 lines.

michikrug commented:
Ah, I see. Thanks for the quick update.

What about the gym webhook message?

Obihoernchen commented:
Tested pokemon scanning, encountering and pokestops + gym + gym-info.
Didn't test tutorial yet.

Some early test results:
- ~15% less memory usage
- slightly lower CPU usage (5% maybe)

friscoMad commented:
@michikrug I don't see any problem with gym webhooks, there are 2 wh for gyms:

- The hook with the info from parsemap that can be disabled with 'webhook-updates-only' it seens all the visible info but not the full pokeon rooster and defending players.
-   The hook with the full gim info, it will only get this info is --gym-info is enabled.

Please tell me which one is missing or the kind of problem you have.

michikrug commented:
I have set "webhook-updates-only" and therefore it should relate to the "gym_details" webhook messages. At least before the merge PokeAlarm received the messages.

AEtHeLsYn commented:
@friscoMad @michikrug According to command line documentation:
```
--webhook-updates-only
                    Only send updates (Pokemon & lured pokestops). [env
                    var: POGOMAP_WEBHOOK_UPDATES_ONLY]
```

If the switch is enabled no gym update info is sent to the webhook

michikrug commented:
@AEtHeLsYn This can't be true, since it was working for months. Also the code, afaik, tells other.

michikrug commented on pogom/models.py:
```diff
@@ -2354,91 +2347,69 @@ def parse_gyms(args, gym_responses, wh_update_queue, db_update_queue):
-                        'individual_attack', 0),
-                    'trainer_name': member['trainer_public_profile']['name'],
-                    'trainer_level': member['trainer_public_profile']['level'],
+                wh_gym_pokemon = gym_pokemon[i].copy()
```
This part somehow stops correct gym_detail webhook messages to be sent. I removed it for testing and it worked again.

sebastienvercammen replied:
@friscoMad @Alderon86 Has someone reviewed this feedback?

friscoMad replied:
Yup this was addressed at 72b40e8

friscoMad commented:
Fixed issue with webhooks and gym detail, there was a silent error when converting gym details to json due to the last_seen field that was added when changed to code to copy.

sebastienvercammen commented on pogom/account.py:
```diff
@@ -268,39 +269,35 @@ def tutorial_pokestop_spin(api, player_level, forts, step_location, account):
+    if in_radius((fort.latitude, fort.longitude), step_location,
                  spinning_radius):
-        log.debug('Attempt to spin Pokestop (ID %s)', fort['id'])
+        log.debug('Attempt to spin Pokestop (ID %s)', fort.id)
```
Missing dot in log message.

sebastienvercammen commented on pogom/models.py:
```diff
@@ -1789,8 +1789,8 @@ def parse_map(args, map_dict, step_location, db_update_queue, wh_update_queue,
-    cells = map_dict['responses']['GET_MAP_OBJECTS']['map_cells']
-    # Get the level for the pokestop spin, and to send to webhook.
+    cells = map_dict['responses']['GET_MAP_OBJECTS'].map_cells
+    # Get the level for the pokestop spin in any case delete inventory
```
Punctuation.

sebastienvercammen commented on pogom/models.py:
```diff
@@ -2354,91 +2347,71 @@ def parse_gyms(args, gym_responses, wh_update_queue, db_update_queue):
+                    'trainer_name': member.trainer_public_profile.name,
+                    'trainer_level': member.trainer_public_profile.level
                 })
+                # Otherwise json conversion will fail and nothing will be sent
```
Punctuation.

friscoMad commented:
Rebased and fixed punctuation as requested, feedback about gym info problem was addressed some time ago there was a silent exception when converting last seen date to json, so I just removed it as it was not being sent before.

urinella commented:
`Exception while downloading map: TypeError('call() takes exactly 1 argument (2 given)',)`

I get this error from last version @friscoMad

friscoMad commented:
At what line? I think you did merge this pr into something else as all calls are passing only False in every call. 
Also I am using this branch clean for my map and is running smoothly.

urinella commented:
Using clean branch. Only this PR for testing.
I cannot tell you the line because I only get this error and nothing else.
Maybe I can try to look at verbose later.
```
git fetch RM pull/2026/head:ram
From https://github.com/RocketMap/RocketMap
 * [new ref]         refs/pull/2026/head -> ram
$ git checkout ram
Previous HEAD position was 57e45f6... Update pgoapi to fix json login errors (#2090)
Switched to branch 'ram'
``` 

Btw: scanning Pokemon, Stops and Gyms using -speed.

friscoMad commented:
Ok, the "2 given" part misguided me, you are using an non up to date pgoapi, that's why it is complaining.
Please:
pip uninstall pgoapi
pip install -r requirements.txt --upgrade

urinella commented:
I'm on latest pgoapi... 78e79bda8b78f970d2b3692bd1a5bd17cefa5fdd
Downgrade to 1a77611c19f46d76681c515ddb0c9ab15c002d11 doesn't work, still same error. 
Sorry.
Maybe someone else can test the latest version of the PR. Could be my server, I don't know.

friscoMad commented:
@urinella That's really strange, I am at  1a77611c19f46d76681c515ddb0c9ab15c002d11 and working will try with latest version.
Please if you can contact me on discord (friscoMad) so we can debug this issue, if you can't please post your full ```pip list```, then do ```pip show pgoapi``` go to the location shown there + "/pgopai" and do a ```cat pgoapi.py | grep use_``` if should show something like ```def call(self, use_dict = True):``` if it does not show anything then for some reason you are on an older pgoapi

friscoMad commented:
Just to make it clear, @urinella was having issues with a second old pgoapi installed in a different directory, when he did fix that the errors were gone. So it was not a problem of this PR.
